okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,256 @@
+                            + "the number of configured backoffs are exhausted", exception));
+                    return null;
+                }
+
+                Duration backoff = exception.getRetryAfter().orElse(nonAdvertizedBackoff.get());
+                log.debug("Rescheduling call after backoff", SafeArg.of("backoffMillis", backoff.toMillis()));
+                scheduleExecution(callback, backoff);
+                return null;
+            }
+
+            @Override
+            public Void visit(QosException.RetryOther exception) {
+                if (maxNumRelocations <= 0) {
+                    callback.onFailure(call,
+                            new IOException("Exceeded the maximum number of allowed "
+                                    + "redirects for initial URL: " + call.request().url()));
+                } else {
+                    Optional<HttpUrl> redirectTo = urls.redirectTo(request().url(),
+                            exception.getRedirectTo().toString());
+                    if (!redirectTo.isPresent()) {
+                        callback.onFailure(call,
+                                new IOException("Failed to determine valid redirect URL "
+                                        + "for '" + exception.getRedirectTo() + "' and base URLs "
+                                        + urls.getBaseUrls()));
+                    } else {
+                        Request redirectedRequest = request().newBuilder()
+                                .url(redirectTo.get())
+                                .build();
+                        client.newCallWithMutableState(
+                                redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                .enqueue(callback);
+                    }
+                }
+                return null;
+            }
+
+            @Override
+            public Void visit(QosException.Unavailable exception) {
+                Optional<Duration> backoff = backoffStrategy.nextBackoff();
+                if (!backoff.isPresent()) {
+                    log.debug("Max number of retries exceeded, failing call");
+                    callback.onFailure(call,
+                            new IOException("Failed to complete the request due to a "
+                                    + "server-side QoS condition: 503", exception));
+                } else {
+                    log.debug("Rescheduling call after backoff",
+                            SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                    scheduleExecution(callback, backoff.get());
+                }
+                return null;
+            }
+        };
+    }
+
+
+    // TODO(rfink): Consider removing RemotingOkHttpCall#doClone method, #627
+    @Override
+    public RemotingOkHttpCall doClone() {
+        return new RemotingOkHttpCall(getDelegate().clone(), backoffStrategy, urls, client, schedulingExecutor,
```

Comment:
If `backoffStrategy` has interior mutability, I don't think it's really fair to call this `clone` anymore because if you clone `callA` to get `callB` then they'd both share the same mutable `BackoffStrategy` instance - calls from either instance (presumably on different threads) would then mutate the same object.  This seems kinda antithetical to the normal 'clone' contract.

Reply:
i think this is compatible with the Call contract:   
```
/**
   * Create a new, identical call to this one which can be enqueued or executed even if this call
   * has already been.
   */
```
, as well as with a general clone contract for mutable objects.

added some javadoc

Reply:
OK sounds like we're not aligned on this one, personally I'd write a bit more code to eliminate the last little bit of mutability and redo that BackoffStrategy so that it's immutable, but it's your PR not mine.

Reply:
where would you prefer to keep state that is mutable across retries?

Reply:
If we reframed the concept of backoff handling to be capture by:

```
interface BackoffSupplier {
    Optional<Duration> currentBackoffDuration(); //doesn't mutate anything
    BackoffSupplier increaseBackoff(); // constructs and returns a new instance
}
```

When you call `client.newCallWithMutableState` you could then pass in `backoff.increaseBackoff()`.

Reply:
Don't think that's better since it means that you always need to remember to call two methods on the backoff object.